globus_action_provider_tools/data_types.py

```diff
@@ -64,6 +64,20 @@ class ActionProviderDescription(BaseModel):
     event_types: Optional[List[EventType]] = None
 
 
+class RequestObject(BaseModel):
+    """Use pydantic to enforce that the request object is a Python dictionary."""
+
+    request_object: Dict[str, Any]
+
+    @validator("request_object", pre=True)
+    def validate_request_object(cls, request_object):
+        if not isinstance(request_object, dict):
+            raise TypeError("the request object must be a JSON object")
+        if not all(isinstance(k, str) for k in request_object):
+            raise TypeError("all top-level keys in the request object must be strings")
+        return request_object
```

Comment:
All of this validation fits into vanilla pydantic validation
Relevant Features:
  * [Custom Root Types](https://pydantic-docs.helpmanual.io/usage/models/#custom-root-types)
  * [Strict Types](https://pydantic-docs.helpmanual.io/usage/types/#strict-types) (otherwise pydantic will auto-coerce things like ints into strings)

Definition
```python
>>> import typing as t
>>> from pydantic import BaseModel, StrictStr

>>> class MyModel(BaseModel):
...    __root__: t.Dict[StrictStr, t.Any]
```

Usage
```python
# Normal case
>>> MyModel.parse_obj({"a": "b", "b": {"c": ["d", "e"]}}) 
MyModel(__root__={'a': 'b', 'b': {'c': ['d', 'e']}})


# Non-json-obj
>>> MyModel.parse_obj("abcd")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pydantic/main.py", line 521, in pydantic.main.BaseModel.parse_obj
  File "pydantic/main.py", line 341, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 1 validation error for MyModel
__root__
  value is not a valid dict (type=type_error.dict)

>>> MyModel.parse_obj(["a", "b"]) 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pydantic/main.py", line 521, in pydantic.main.BaseModel.parse_obj
  File "pydantic/main.py", line 341, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 1 validation error for MyModel
__root__
  value is not a valid dict (type=type_error.dict)

# Non-str keys
>>> MyModel.parse_obj({12: "b", "a": "b", "b": {"c": ["d", "e"]}}) 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pydantic/main.py", line 521, in pydantic.main.BaseModel.parse_obj
  File "pydantic/main.py", line 341, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 1 validation error for MyModel
__root__ -> __key__
  str type expected (type=type_error.str)
```

Reply:
Doggone, this is what I was looking for. FANTASTIC! I'll switch this out immediately.